src/hidimstat/distilled_conditional_randomization_test.py

```diff
@@ -213,27 +235,45 @@ def fit(self, X, y):
                     screening_threshold=self.screening_threshold,
                     random_state=rng,
                 )
-                if self.refit:
-                    self.estimator.fit(X_[:, self.selection_set_], y_)
         else:
             self.selection_set_ = np.ones(X_.shape[1], dtype=bool)
 
+        # Refit the model on the selected features if required or if no estimated
+        # coefficients were provided and screening was not performed
+        if self.refit or (
+            (self.screening_threshold is None) and self.estimated_coef is None
+        ):
+            self.estimator.fit(X_[:, self.selection_set_], y_)
+        elif (self.screening_threshold is not None) and (self.estimated_coef is None):
+            self.estimator = lasso_model_
+
         if self.estimated_coef is not None:
             self.coefficient_ = self.estimated_coef
+            self.intercept_ = (
+                self.estimated_intercept if self.estimated_intercept is not None else 0
+            )
         elif self.reuse_screening_model and (self.screening_threshold is not None):
-            self.coefficient_ = lasso_model_.coef_
+            # Flatten to handle logistic regression case
+            self.coefficient_ = lasso_model_.coef_.flatten()
+            self.intercept_ = lasso_model_.intercept_
             # optimization to reduce the number of elements different to zeros
             self.coefficient_[~self.selection_set_] = 0
         else:
-            self.estimator.fit(X_[:, self.selection_set_], y_)
+            # If the model is linear, store the coefficients
             if hasattr(self.estimator, "coef_"):
-                self.coefficient_ = self.estimator.coef_
-                # optimization to reduce the number of elements different to zeros
-                # TODO: the multi-dimensional indexing should be removed when the
-                # classification is properly handled
-                self.coefficient_[..., ~self.selection_set_] = 0
+                self.coefficient_ = np.zeros(X.shape[1])
+                self.coefficient_[self.selection_set_] = self.estimator.coef_.flatten()
+                self.estimator.coef_ = self.coefficient_
+                self.intercept_ = self.estimator.intercept_
             else:
                 self.coefficient_ = None
+        # Save sample weights that will be used for fitting the X-distillation and
+        # computing the Fisher information matrix
+        if self.is_logistic_:
+            self.lasso_weights_ = (
```

Comment:
If it's possible, can you add the reference to the equation for this calculus?

Reply:
done